src/openloco/windows/CompanyWindow.cpp

```diff
@@ -0,0 +1,1501 @@
+#include "../ui/dropdown.h"
+#include "../viewportmgr.h"
+#include "../widget.h"
+
+using namespace openloco::interop;
+
+namespace openloco::ui::windows::CompanyWindow
+{
+    loco_global<uint16_t[8], 0x112C826> _common_format_args;
+
+    namespace common
+    {
+        enum widx
+        {
+            frame,
+            caption,
+            close_button,
+            panel,
+            tab_status,
+            tab_details,
+            tab_colour_scheme,
+            tab_finances,
+            tab_cargo_delivered,
+            tab_challenge,
+            company_select,
+        };
+
+        const uint64_t enabledWidgets = (1 << widx::caption) | (1 << widx::close_button) | (1 << widx::tab_status) | (1 << widx::tab_details) | (1 << widx::tab_colour_scheme) | (1 << widx::tab_finances) | (1 << widx::tab_cargo_delivered) | (1 << widx::tab_challenge) | (1 << widx::company_select);
+
+#define commonWidgets(frameWidth, frameHeight, windowCaptionId)                                                                                \
+    make_widget({ 0, 0 }, { frameWidth, frameHeight }, widget_type::frame, 0),                                                                 \
+        make_widget({ 1, 1 }, { frameWidth - 2, 13 }, widget_type::caption_24, 0, windowCaptionId),                                            \
+        make_widget({ frameWidth - 15, 2 }, { 13, 13 }, widget_type::wt_9, 0, image_ids::close_button, string_ids::tooltip_close_window),      \
+        make_widget({ 0, 41 }, { frameWidth, 120 }, widget_type::panel, 1),                                                                    \
+        make_remap_widget({ 3, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_company_owner_and_status),          \
+        make_remap_widget({ 34, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_company_headquarters_and_details), \
+        make_remap_widget({ 65, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_company_colour_scheme),            \
+        make_remap_widget({ 96, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_company_finances),                 \
+        make_remap_widget({ 127, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_cargo_delivered),                 \
+        make_remap_widget({ 158, 15 }, { 31, 27 }, widget_type::wt_8, 1, image_ids::tab, string_ids::tooltip_company_challenge_for_this_game), \
+        make_widget({ 0, 14 }, { 26, 26 }, widget_type::wt_9, 0, string_ids::null, string_ids::tooltip_select_company)
+
+        // 0x004343FC
+        static void toggleRenameCompany(window* self)
+        {
+            self->disabled_widgets = 0;
+            if (self->number != companymgr::get_controlling_id())
+                self->disabled_widgets |= (1 << widx::tab_challenge);
+        }
+
+        // 0x00498E9B
+        static void enableRenameByCaption(window* self)
```

Comment:
This is the town window version. The company window uses `0x431E9B`.

Reply:
The address comment needs adjusting, yes.

Reply:
0x431E9B also checks if the player is the owner. I think this would also be more clear if not-using inverse statements. e.g.

```c
if(is_editor() || is_owner()) {
   self->enabled_widgets |= (1 << caption);
} else {
   self->enabled_widgets &= ~(1 << caption);
}
```